src/routing/middleware/validate.rs

```diff
@@ -41,6 +43,16 @@ pub async fn middleware(
             .map_err(|e| ErrorCause::CanisterIdIncorrect(e.to_string()))?
     }
 
+    if state.canister_id_from_referer && lookup.canister_id.is_none() {
+        lookup.canister_id = match canister_id_from_referer_host(state.resolver, &request)
```

Comment:
I also think maybe we just need to return `Option<Principal>` from the functions that resolve it, without `Result`. This way we can just chain them using `.or_else(|| ...)` and then just fail in the end if we can't extract anything.

Btw this way it was done back in the days in `icx-proxy`.